### ISSUE TITLE
How many iterations is a sample be trained ?

### ISSUE THREAD

**Issue by wqshmzh** [snumprlab/cl-alfred#11]
The default value of args.online_iter is 1, which means that each sample is trained in only 1 iteration. However, the calling of self.online_train() in cama.py set iterations and stream_batch_size as int(self.num_updates) and self.temp_batchsize as below:
 info, logits = self.online_train(
                self.temp_batch,
                self.batch_size,
                iterations=int(self.num_updates),
                stream_batch_size=self.temp_batchsize
            )
Now a streamed batch has self.temp_batchsize samples (ignores the samples collected from memory) and is trained in int(self.num_updates) times and the overal iterations is self.temp_batchsize*int(self.num_updates), which does not equal to self.temp_batchsize.
How many iterations do you exactly set to train each sample ?

**Comment by dbd05088** [comment]
As mentioned earlier, we train the model with self.online_iter batch iterations per sample encountered.
Since our framework assumes that samples are encountered one by one, we need to accumulate streaming samples to form a training batch in ER-based methods (e.g., ER, DER, and CAMA), where these batches are composed of both streaming data and memory data.

Let's assume the training batch size is 16, and self.online_iter = 1. 
Since ER create training batches with half stream data and half memory data (self.temp_batchsize = 8 and self.memory_batchsize = 8), we need to accumulate 8 stream samples. 
After this accumulation, self.num_updates = 8 (because self.num_updates += self.online_iter for each sample encountered). Therefore, the model undergoes 8 batch iterations of training.
Here, accumulating 8 stream samples and training the model with 8 batch iterations after accumulation means 1 batch iteration per stream sample encountered.

**Comment by wqshmzh** [comment]
However, I find that the code collects a batch containing 8 streaming samples first (self.memory.get_data(16, 8) in cama.py returns a batch containing 8 streaming samples), and then feeds the features of the 8 streaming samples in this batch to self.model.forward(feat) (In the forward() func in vnn.py, the self.step() func processes the features of 8 instead of 1 steaming samples at a time) for 8 times , which means that each streaming sample in this batch is iterated for 8 times.
Also, why does args.temp_batchsize equal to args.batchsize - 2 * args.batchsize // 3 in CAMA ?

**Comment by bhkim94** [comment]
@dbd05088 Do you have any opinions on this?

**Comment by dbd05088** [comment]
We haven't mentioned **"1 streaming sample at a time"**. As previously noted, we perform **"one batch iteration per stream sample encountered"**. Moreover, to balance memory data and current stream data, we collect stream samples.

Regarding the args.temp_batchsize = args.batchsize - 2 * args.batchsize // 3 in CAMA, we followed the implementation details of DER. The rationale for this choice is explained in the pseudocode of the DER paper. Note that CAMA is a modified version of DER, thus we followed the hyperparameter of the original DER implementation.

**Comment by bhkim94** [comment]
Closing this due to inactivity. Feel free to reopen this if you have any concerns.

**Comment by wqshmzh** [comment]
> We haven't mentioned **"1 streaming sample at a time"**. As previously noted, we perform **"one batch iteration per stream sample encountered"**. Moreover, to balance memory data and current stream data, we collect stream samples.
> 
> Regarding the args.temp_batchsize = args.batchsize - 2 * args.batchsize // 3 in CAMA, we followed the implementation details of DER. The rationale for this choice is explained in the pseudocode of the DER paper. Note that CAMA is a modified version of DER, thus we followed the hyperparameter of the original DER implementation.

Do you mean that the batch containing 8 streaming samples and 8 memory samples is iterated for 8 times? 
However, the seq2seq model doesn't update itself with the 8 streaming samples one by one. It uses all the 8 streaming samples at once to update itself. So in this way, each streaming sample in this batch is iterated for 8 times.

I am confused about this. If each streaming sample is supposed to be iterated for 1 time, then the batch is supposed to be iterated for 1 time too, because the seq2seq model processes all the streaming samples in this batch for one model update. Only when the seq2seq model updates itself with 1 streaming sample in this batch would result in each sample iterating for 1 time.
